hypothesis-python/src/hypothesis/internal/conjecture/utils.py

```diff
@@ -178,88 +178,118 @@ def boolean(data):
+            # given by the fraction of that interval that was <= our previous
+            # p.
+
+            # We then take advantage of the fact that we have control of the
+            # labelling to make this shrink better, using the following tricks:
+
+            # If p is <= 0 or >= 1 the result of this coin is certain. We make
+            # sure to write to the data stream anyway so that these don't cause
+            # difficulties when shrinking. Note that we ensured that this can't
+            # happen on the first iteration of the loop, but there's nothing to
+            # stop it happening on a later one.
+            p = self.probability
+            opts = 2 ** self.n_bits
+            falsey = floor(opts * (1 - p))
+            truthy = floor(opts * p)
+            if p <= 0:
+                self.rig(data, False)
                 result = False
-            elif i <= 1:
-                # We special case so that zero is always false and 1 is always
-                # true which makes shrinking easier because we can always
-                # replace a truthy block with 1. This has the slightly weird
-                # property that shrinking from 2 to 1 can cause the result to
-                # grow, but the shrinker always tries 0 and 1 first anyway, so
-                # this will usually be fine.
-                result = bool(i)
+            elif p >= 1:
+                self.rig(data, True)
+                result = True
             else:
-                # Originally everything in the region 0 <= i < falsey was false
-                # and everything above was true. We swapped one truthy element
-                # into this region, so the region becomes 0 <= i <= falsey
-                # except for i = 1. We know i > 1 here, so the test for truth
-                # becomes i > falsey.
-                result = i > falsey
-        break
-    data.stop_example()
-    return result
+                remainder = opts * p - truthy
+
+                if falsey + truthy == opts:
+                    if isinstance(p, Fraction):
+                        m = p.numerator
+                        n = p.denominator
+                    else:
+                        m, n = p.as_integer_ratio()
+                    assert n & (n - 1) == 0, n  # n is a power of 2
+                    assert n > m > 0
+                    truthy = m
+                    falsey = n - m
+                    partial = False
+                else:
+                    partial = True
+
+                i = data.draw_bits(self.n_bits)
+
+                # We always label the region that causes us to repeat the loop
+                # is opts - 1 so that shrinking this byte never causes us to
+                # need to draw more data.
+                if partial and i + 1 == opts:
+                    p = remainder
+                    continue
+                if i <= 1:
+                    # We special case so that zero is always false and 1 is
```

Comment:
`We special case so...` - sounds a bit strange, at least for me. Is that a typo?

Reply:
I don't think it's strictly wrong, but you're right that it reads oddly, thanks. Fixed now.